### PR TITLE
fix: Clear cache after update policies.

### PR DESCRIPTION
**Pull request by 00LT00**
Signed-off-by: 00LT00 <lightning@zerokirin.online>

#568

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece @GopherJ @dovics please review.

File: enforcer_cached.go [hunk 1]
```diff
@@ -22,9 +22,10 @@ import (
 // CachedEnforcer wraps Enforcer and provides decision cache
 type CachedEnforcer struct {
 	*Enforcer
-	m           map[string]bool
-	enableCache bool
-	locker      *sync.RWMutex
+	m              map[string]bool
+	enableCache    bool
```
Comment (GopherJ) [inline enforcer_cached.go:26]: `enableCache` is not necessary because cache must be cleaned

File: enforcer_cached.go [hunk 2]
```diff
@@ -47,6 +49,11 @@ func (e *CachedEnforcer) EnableCache(enableCache bool) {
 	e.enableCache = enableCache
 }
 
+// AutoClear determines whether to clear cache after any operation for policy. if AutoClear is true, cache will be clear after update policy.
+func (e *CachedEnforcer) AutoClearCache(autoClearCache bool) {
```
Comment (GopherJ) [inline enforcer_cached.go:53]: not necessary

Comment (00LT00) [reply enforcer_cached.go:53]: I think there may be users who are used to the way. Without this, will some users be troubled?

Comment (nodece) [reply enforcer_cached.go:53]: when policy changes, we must clear cache, so `AutoClearCache`  is unnecessary.

Comment (00LT00) [reply enforcer_cached.go:53]: ok

**Comment by 00LT00** [comment]
plz review @GopherJ @nodece

File: enforcer_cached.go [hunk 3]
```diff
@@ -93,3 +93,36 @@ func (e *CachedEnforcer) setCachedResult(key string, res bool) {
...
+
+// RemovePolicy removes an authorization rule from the current policy.
+func (e *CachedEnforcer) RemovePolicy(params ...interface{}) (bool, error) {
+	ok, err := e.RemoveNamedPolicy("p", params...)
+	e.InvalidateCache()
+	return ok, err
+}
+
+// RemovePolicies removes authorization rules from the current policy.
+func (e *CachedEnforcer) RemovePolicies(rules [][]string) (bool, error) {
```
Comment (GopherJ) [inline enforcer_cached.go:124]: How about `RemoveFilteredPolicy`?

**Review by GopherJ — APPROVED** [review]

File: enforcer_cached.go [hunk 4]
```diff
@@ -93,3 +93,36 @@ func (e *CachedEnforcer) setCachedResult(key string, res bool) {
...
 	e.m = make(map[string]bool)
 }
+
+/* Clear cache after update Policy*/
+// AddPolicy adds an authorization rule to the current policy.
+// If the rule already exists, the function returns false and the rule will not be added.
+// Otherwise the function returns true by adding the new rule.
+func (e *CachedEnforcer) AddPolicy(params ...interface{}) (bool, error) {
+	ok, err := e.AddNamedPolicy("p", params...)
+	e.InvalidateCache()
```
Comment (GopherJ) [inline enforcer_cached.go:103]: also please check `if ok`, avoid unecessary clear is important

File: enforcer_cached.go [hunk 5]
```diff
@@ -93,3 +93,36 @@ func (e *CachedEnforcer) setCachedResult(key string, res bool) {
...
+	return ok, err
+}
+
+// AddPolicies adds authorization rules to the current policy.
+// If the rule already exists, the function returns false for the corresponding rule and the rule will not be added.
+// Otherwise the function returns true for the corresponding rule by adding the new rule.
+func (e *CachedEnforcer) AddPolicies(rules [][]string) (bool, error) {
+	ok, err := e.AddNamedPolicies("p", rules)
```
Comment (GopherJ) [inline enforcer_cached.go:111]: check pls `ok`

File: enforcer_cached.go [hunk 6]
```diff
@@ -93,3 +93,36 @@ func (e *CachedEnforcer) setCachedResult(key string, res bool) {
...
+	e.InvalidateCache()
+	return ok, err
+}
+
+// RemovePolicy removes an authorization rule from the current policy.
+func (e *CachedEnforcer) RemovePolicy(params ...interface{}) (bool, error) {
+	ok, err := e.RemoveNamedPolicy("p", params...)
```
Comment (GopherJ) [inline enforcer_cached.go:118]: check pls `ok`

File: enforcer_cached.go [hunk 7]
```diff
@@ -93,3 +93,36 @@ func (e *CachedEnforcer) setCachedResult(key string, res bool) {
...
+	e.InvalidateCache()
+	return ok, err
+}
+
+// RemovePolicies removes authorization rules from the current policy.
+func (e *CachedEnforcer) RemovePolicies(rules [][]string) (bool, error) {
+	ok, err := e.RemoveNamedPolicies("p", rules)
```
Comment (GopherJ) [inline enforcer_cached.go:125]: check pls `ok`

File: enforcer_cached.go (hunk 4 above)
Comment (00LT00) [reply enforcer_cached.go:103]: sure

File: enforcer_cached.go (hunk 3 above)
Comment (00LT00) [reply enforcer_cached.go:124]: I don't quite understand. Does it mean to add this function?

Comment (GopherJ) [reply enforcer_cached.go:130]: you need to clear cache in RmoveFilteredPolicy

**Comment by GopherJ** [comment]
I think this PR won't solve problems, because it works only when users call

1. AddPolicy
2. AddPolicies
3. RemovePolicy
4. RemovePolicies

how can we ensure it overwrites all the management/rbac apis? 

if users call e. AddRoleForUser, it doesn't find this function in cachedEnforcer 

so it'll still call EEnforcer's function

**Comment by GopherJ** [comment]
See maybe how casbin-cpp and casbin-rs solved this issue

**Comment by 00LT00** [comment]
ok

**Comment by 00LT00** [comment]
But where the casbin-cpp's code. I see it doesn't clear cache after change policy. Or do you mean that cpp is also facing this problem now?

image

And the test is:
image

**Comment by GopherJ** [comment]
@divy9881 @nodece Any idea on this?

**Pull request closed without merging**